File: day9/python/main.py
from copy import deepcopy
from math import sqrt


def draw(nodes):
    """
    Draws the rope in a grid_size x grid_size grid, 
    highest number shown if nodes are on top of eachother
    """
    grid_size = 50
    row = ['.'] * grid_size
    grid = [deepcopy(row) for i in range(grid_size)]
    node_num = 0
    origin = (int(grid_size / 2) - 1)
    grid[origin][origin] = "s"

    for node in nodes:
        x = node[0]
        y = node[1]
        x = x + (int(grid_size / 2) - 1)
        y = -y + (int(grid_size / 2) - 1)
        grid[y][x] = str(node_num)
        node_num += 1

    print("-" * (grid_size + 5))
    row_num = 1
    for row in grid:
        print(f"{row_num:2}", end="")
        row_num += 1
        for char in row:
            print(char, end="")
        print(end='\n')
    print("-" * (grid_size + 5))


def distance(node0, node1):
    [x0, y0] = node0
    [x1, y1] = node1
    return sqrt((x0 - x1) ** 2 + (y0 - y1) ** 2)


def move_other_nodes(nodes):
    """ Moves the other nodes according to the movement of head """
    for i in range(len(nodes) - 1):
        head = nodes[i]
        tail = nodes[i + 1]
        [x0, y0] = head
        [x1, y1] = tail
        # actual distance limit is sqrt(2) but next step would be sqrt(5) so using 2 is fine.
        if distance(head, tail) >= 2:
            # Head and tail are too far away, find out which way and move accordingly
            if x0 == x1:  # On the same row, which means 2 steps Up or Down
                if y0 > y1:
                    y1 += 1
                else:
                    y1 -= 1
            elif y0 == y1:  # On the same columns, which means 2 steps Left or Right
                if x0 > x1:
                    x1 += 1
                else:
                    x1 -= 1
            else:  # Not touching, must move diagonally
                if y0 > y1:
                    y1 += 1
                else:
                    y1 -= 1
                if x0 > x1:
                    x1 += 1
                else:
                    x1 -= 1
            tail[0] = x1
            tail[1] = y1


def move(nodes, direction):
    """ Moves the head according to direction and then moves other nodes accordingly"""
    head = nodes[0]
    match direction:
        case "U":
            head[1] += 1
        case "D":
            head[1] -= 1
        case "L":
            head[0] -= 1
        case "R":
            head[0] += 1
    move_other_nodes(nodes)


def first_task(lines):
    nodes = [[0, 0], [0, 0]]
    unique_tails = set()
    for line in lines:
        direction = line[0]
        times = int(line[2:])
        while times:
            times -= 1
            move(nodes, direction)
            unique_tails.add(tuple(nodes[-1]))
    print(f"first_task: {len(unique_tails)}")


def second_task(lines):
    nodes = [[0, 0] for _ in range(10)]
    unique_tails = set()
    for line in lines:
        direction = line[0]
        times = int(line[2:])
        while times:
            times -= 1
            move(nodes, direction)
            unique_tails.add(tuple(nodes[-1]))
    print(f"second_task: {len(unique_tails)}")


def main():
    f = open("input.txt", "r")
    lines = f.readlines()
    first_task(lines)
    second_task(lines)


if __name__ == "__main__":
    main()

File: day9/rust/src/main.rs
use std::collections::HashSet;
use std::fs;

fn distance(node0: (i32, i32), node1: (i32, i32)) -> f32 {
    f32::sqrt((node0.0 - node1.0).pow(2) as f32 + (node0.1 - node1.1).pow(2) as f32)
}

fn move_other_nodes(nodes: &mut [(i32, i32)]) {
    for i in 0..nodes.len() - 1 {
        let x0 = nodes[i].0;
        let y0 = nodes[i].1;
        let mut x1 = nodes[i + 1].0;
        let mut y1 = nodes[i + 1].1;
        if distance(nodes[i], nodes[i + 1]) >= 2.0 {
            if x0 == x1 {
                if y0 > y1 {
                    y1 += 1;
                } else {
                    y1 -= 1;
                }
            } else if y0 == y1 {
                if x0 > x1 {
                    x1 += 1;
                } else {
                    x1 -= 1;
                }
            } else {
                if y0 > y1 {
                    y1 += 1;
                } else {
                    y1 -= 1;
                }
                if x0 > x1 {
                    x1 += 1;
                } else {
                    x1 -= 1;
                }
            }
        }
        nodes[i + 1].0 = x1;
        nodes[i + 1].1 = y1;
    }
}

fn move_head(nodes: &mut [(i32, i32)], direction: char) {
    let head: &mut (i32, i32) = nodes.get_mut(0).unwrap();
    match direction {
        'U' => head.1 += 1,
        'D' => head.1 -= 1,
        'L' => head.0 -= 1,
        'R' => head.0 += 1,
        _ => {}
    }
    move_other_nodes(nodes);
}

fn first_task(moves: Vec<(char, i32)>) {
    let mut nodes = vec![(0, 0), (0, 0)];
    let mut unique_tails = HashSet::<(i32, i32)>::new();
    for mv in moves {
        let direction = mv.0;
        let mut times = mv.1;
        while times > 0 {
            times -= 1;
            move_head(nodes.as_mut_slice(), direction);
            unique_tails.insert(*nodes.last().unwrap());
        }
    }
    println!("first_task: {}", unique_tails.len());
}

fn second_task(moves: Vec<(char, i32)>) {
    let mut nodes = vec![
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
        (0, 0),
    ];
    let mut unique_tails = HashSet::<(i32, i32)>::new();
    for mv in moves {
        let direction = mv.0;
        let mut times = mv.1;
        while times > 0 {
            times -= 1;
            move_head(nodes.as_mut_slice(), direction);
            unique_tails.insert(*nodes.last().unwrap());
        }
    }
    println!("second_task: {}", unique_tails.len());
}

fn main() {
    let file_content = fs::read_to_string("../input.txt").unwrap();
    let lines = file_content.split('\n');
    let mut moves: Vec<(char, i32)> = vec![];
    for line in lines {
        if line.is_empty() {
            break;
        }
        let mut split = line.split(' ');
        moves.push((
            split.next().unwrap().chars().next().unwrap(),
            split.next().unwrap().parse().unwrap(),
        ));
    }
    first_task(moves.clone());
    second_task(moves);
}
